File: scripts/collector_x_apify.py
#!/usr/bin/env python3
import os
import sys
import requests
import json
import time

APIFY_API_URL = "https://api.apify.com/v2"

def get_token():
    if os.path.exists(".env"):
        with open(".env") as f:
            for line in f:
                if line.startswith("APIFY_API_TOKEN="):
                    return line.strip().split("=")[1]
    return os.getenv("APIFY_API_TOKEN")

def run_twitter_scraper(token, query, max_tweets=10):
    # Using 'apify~twitter-scraper'
    actor_id = "apify~twitter-scraper"
    
    run_input = {
        "searchQueries": [query],
        "maxTweets": max_tweets,
        "addUserInfo": True
    }
    
    print(f"Starting Apify Actor {actor_id} for query: {query}...")
    headers = {"Content-Type": "application/json"}
    start_url = f"{APIFY_API_URL}/acts/{actor_id}/runs?token={token}"
    
    res = requests.post(start_url, json=run_input, headers=headers)
    run_data = res.json()
    
    if "error" in run_data:
        print(f"Error starting Actor: {run_data['error']}")
        return None
    
    run_id = run_data["data"]["id"]
    dataset_id = run_data["data"]["defaultDatasetId"]
    
    print(f"Run started: {run_id}. Waiting for completion...")
    
    # Poll for completion
    while True:
        status_url = f"{APIFY_API_URL}/acts/{actor_id}/runs/{run_id}?token={token}"
        status_res = requests.get(status_url)
        status_data = status_res.json()
        
        status = status_data["data"]["status"]
        if status == "SUCCEEDED":
            break
        elif status in ["FAILED", "ABORTED", "TIMED-OUT"]:
            print(f"Actor run failed with status: {status}")
            return None
        
        time.sleep(10)
    
    # Fetch results
    print("Fetching results from dataset...")
    items_url = f"{APIFY_API_URL}/datasets/{dataset_id}/items?token={token}"
    items_res = requests.get(items_url)
    return items_res.json()

if __name__ == "__main__":
    token = get_token()
    if not token:
        print("Error: APIFY_API_TOKEN not found.")
        sys.exit(1)
    
    results = run_twitter_scraper(token, "AI Agents 2026")
    if results:
        print(f"SUCCESS: Collected {len(results)} tweets.")
        with open("collected_tweets_sample.json", "w") as f:
            json.dump(results, f, indent=2, ensure_ascii=False)
        print("Sample data saved to collected_tweets_sample.json")

File: scripts/collector_x_apify.js
#!/usr/bin/env node
import fs from 'node:fs';
import axios from 'axios';
import dotenv from 'dotenv';

dotenv.config();

const APIFY_API_URL = "https://api.apify.com/v2";

async function get_token() {
    return process.env.APIFY_API_TOKEN;
}

/**
 * X (Twitter) Data Collector - Full Coverage Edition
 * Modes:
 * 1. 'search'   - Query based (Advanced search supported)
 * 2. 'user'     - Fetch tweets from specific handles
 * 3. 'list'     - Fetch context from a specific X List (URL required)
 * 4. 'profile'  - Fetch metadata for specific handles
 * 5. 'url'      - Direct URL scraping (specific tweets, searches, or profiles)
 */
async function run_twitter_scraper(token, options = {}) {
    const {
        mode = 'search',
        query = '',
        handles = [],
        urls = [], // For 'list' or 'url' modes
        maxTweets = 10,
        actorId = "apify~twitter-scraper-v2"
    } = options;

    console.log(`[X-Collector] Mode: ${mode.toUpperCase()} | Max Items: ${maxTweets}`);

    let runInput = {
        maxTweets: maxTweets,
        addUserInfo: true,
        sort: "Latest" // Default to latest for polling accuracy
    };

    // Advanced Polymorphic Input Mapping
    if (mode === 'search') {
        runInput.searchTerms = Array.isArray(query) ? query : [query];
    } else if (mode === 'user') {
        runInput.twitterHandles = Array.isArray(handles) ? handles : [handles];
    } else if (mode === 'list' || mode === 'url') {
        runInput.startUrls = (Array.isArray(urls) ? urls : [urls]).map(u => ({ url: u }));
    } else if (mode === 'profile') {
        runInput.twitterHandles = Array.isArray(handles) ? handles : [handles];
        runInput.scrapeProfile = true;
    }

    try {
        const runRes = await axios.post(`${APIFY_API_URL}/acts/${actorId}/runs?token=${token}`, runInput);
        const runId = runRes.data.data.id;
        const datasetId = runRes.data.data.defaultDatasetId;

        console.log(`[X-Collector] Run started: ${runId}. Polling latest posts...`);

        // Polling loop
        while (true) {
            const statusRes = await axios.get(`${APIFY_API_URL}/acts/${actorId}/runs/${runId}?token=${token}`);
            const { status } = statusRes.data.data;

            if (status === "SUCCEEDED") break;
            if (["FAILED", "ABORTED", "TIMED-OUT"].includes(status)) {
                throw new Error(`Apify Actor failed with status: ${status}`);
            }
            await new Promise(resolve => setTimeout(resolve, 5000));
        }

        console.log("[X-Collector] Fetching results...");
        const itemsRes = await axios.get(`${APIFY_API_URL}/datasets/${datasetId}/items?token=${token}`);
        return itemsRes.data;

    } catch (error) {
        console.error("[X-Collector] Error:", error.response?.data || error.message);
        return null;
    }
}

async function main() {
    const token = await get_token();
    if (!token) {
        console.error("Error: APIFY_API_TOKEN not found.");
        process.exit(1);
    }

    // Read Mission from config.json
    let config = { mission: { mode: 'search', query: 'AI Agents 2026', maxTweets: 10 } };
    try {
        if (fs.existsSync('config.json')) {
            config = JSON.parse(fs.readFileSync('config.json', 'utf8'));
        }
    } catch (e) {
        console.error("Error reading config.json, using defaults.");
    }

    console.log(`[X-Collector] Starting mission: ${JSON.stringify(config.mission)}`);
    const results = await run_twitter_scraper(token, config.mission);

    if (results) {
        console.log(`[SUCCESS] Collected ${results.length} items.`);
        const filename = `x_data_latest.json`;
        fs.writeFileSync(filename, JSON.stringify(results, null, 2));

        // Update last run in config
        config.settings = config.settings || {};
        config.settings.last_run = new Date().toISOString();
        fs.writeFileSync('config.json', JSON.stringify(config, null, 2));
    }
}

main().catch(err => {
    console.error(err);
    process.exit(1);
});
